crates/shared/src/sources/uniswap_v3/event_fetching.rs

```diff
@@ -0,0 +1,148 @@
+use crate::current_block::RangeInclusive;
+use crate::event_handling::{EventRetrieving, EventStoring};
+use crate::Web3;
+use hex_literal::hex;
+
+use anyhow::{Context, Result};
+use contracts::{
+    uniswap_v3_pool::event_data::{Burn, Mint, Swap},
+    UniswapV3Pool,
+};
+use ethcontract::{
+    common::abi::Error, contract::ParseLog, dyns::DynAllEventsBuilder, errors::ExecutionError,
+    Event, RawLog, H160, H256,
+};
+
+const SWAP_TOPIC: [u8; 32] =
+    hex!("c42079f94a6350d7e6235f29174924f928cc2ac818eb64fed8004e115fbcca67");
+const BURN_TOPIC: [u8; 32] =
+    hex!("0c396cd989a39f4459b5fa1aed6a9a8dcdbc45908acfd67e028cd568da98982c");
+const MINT_TOPIC: [u8; 32] =
+    hex!("7a53080ba414158be7ec69b987b5fb7d07dee101fe85488f0853ae16239d0bde");
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum UniswapV3Event {
+    Burn(Burn),
+    Mint(Mint),
+    Swap(Swap),
+}
+
+impl ParseLog for UniswapV3Event {
+    fn parse_log(log: RawLog) -> Result<Self, ExecutionError> {
+        let standard_event: Option<Result<UniswapV3Event, ExecutionError>> =
+            log.topics.get(0).copied().map(|topic| match topic {
+                H256(BURN_TOPIC) => Ok(UniswapV3Event::Burn(
+                    log.clone().decode(
+                        UniswapV3Pool::raw_contract()
+                            .abi
+                            .event("Burn")
+                            .expect("generated event decode"),
```

Comment:
Not sure about the `expect`s here. We could get bogus data from the node or maybe a weird event we didn't anticipate and this would panic. Would rather return result.

Reply:
This part of the code is copy paste from the example generated by the `contracts` crate build script, since it is the same for all other contracts I would keep it as is, if we want to change it I guess we should change the original build script.

Reply:
> We could get bogus data from the node or maybe a weird event we didn't anticipate and this would panic.

This is expecting that a `Burn` event exists in the ABI. Since the ABI is included at compile time, and the `Burn` event should always be there, I think its fine to expect here.